File: java/src/main/java/org/endeavourhealth/library/model/search/EntityDocument.java
package org.endeavourhealth.library.model.search;

import com.fasterxml.jackson.annotation.JsonSetter;
import org.endeavourhealth.library.model.tripletree.TTIriRef;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;


public class EntityDocument {
  Integer id;
  String iri;
  String name;
  Integer length;
  String preferredName;
  String code;
  String alternativeCode;
  TTIriRef scheme;
  Set<TTIriRef> type = new HashSet<>();
  TTIriRef status;
  Set<SearchTermCode> termCode = new HashSet<>();
  Integer usageTotal;
  String match;
  Set<TTIriRef> isA = new HashSet<>();
  Set<TTIriRef> memberOf = new HashSet<>();
  Integer subsumptionCount;
  Set<String> binding;
  private List<TTIriRef> isDescendentOf = new ArrayList<>();

  public Set<String> getBinding() {
    return binding;
  }

  public EntityDocument setBinding(Set<String> binding) {
    this.binding = binding;
    return this;
  }

  public EntityDocument addBinding(String path, String node) {
    if (this.binding == null) {
      this.binding = new HashSet<>();
    }
    this.binding.add(path + " " + node);
    return this;
  }

  public String getAlternativeCode() {
    return alternativeCode;
  }

  public EntityDocument setAlternativeCode(String alternativeCode) {
    this.alternativeCode = alternativeCode;
    return this;
  }

  public Integer getSubsumptionCount() {
    return subsumptionCount;
  }

  public EntityDocument setSubsumptionCount(Integer subsumptionCount) {
    this.subsumptionCount = subsumptionCount;
    return this;
  }

  public Integer getLength() {
    return length;
  }

  public EntityDocument setLength(Integer length) {
    this.length = length;
    return this;
  }

  public String getPreferredName() {
    return preferredName;
  }

  public EntityDocument setPreferredName(String preferredName) {
    this.preferredName = preferredName;
    return this;
  }

  public Set<TTIriRef> getIsA() {
    return isA;
  }

  public EntityDocument setIsA(Set<TTIriRef> isA) {
    this.isA = isA;
    return this;
  }

  public Integer getId() {
    return id;
  }

  public EntityDocument setId(Integer id) {
    this.id = id;
    return this;
  }

  public String getIri() {
    return iri;
  }

  public EntityDocument setIri(String iri) {
    this.iri = iri;
    return this;
  }

  public String getName() {
    return name;
  }

  public EntityDocument setName(String name) {
    this.name = name;
    return this;
  }

  public String getCode() {
    return code;
  }

  public EntityDocument setCode(String code) {
    this.code = code;
    return this;
  }

  public TTIriRef getScheme() {
    return scheme;
  }

  @JsonSetter
  public EntityDocument setScheme(TTIriRef scheme) {
    this.scheme = scheme;
    return this;
  }

  public TTIriRef getStatus() {
    return status;
  }

  @JsonSetter
  public EntityDocument setStatus(TTIriRef status) {
    this.status = status;
    return this;
  }

  public EntityDocument addType(TTIriRef type) {
    this.type.add(type);
    return this;
  }

  public Integer getUsageTotal() {
    return usageTotal;
  }

  public EntityDocument setUsageTotal(Integer usageTotal) {
    this.usageTotal = usageTotal;
    return this;
  }


  public List<TTIriRef> getIsDescendentOf() {
    return isDescendentOf;
  }

  public EntityDocument setIsDescendentOf(List<TTIriRef> isDescendentOf) {
    this.isDescendentOf = isDescendentOf;
    return this;
  }

  public Set<TTIriRef> getType() {
    return type;
  }

  public EntityDocument setType(Set<TTIriRef> type) {
    this.type = type;
    return this;
  }


  public Set<SearchTermCode> getTermCode() {
    return termCode;
  }

  public EntityDocument setTermCode(Set<SearchTermCode> searchTermCode) {
    this.termCode = searchTermCode;
    return this;
  }

  public EntityDocument addTermCode(String term, String code, TTIriRef status,String keyTerm) {
    SearchTermCode tc = new SearchTermCode();
    tc.setTerm(term).setCode(code).setStatus(status);
    if (term!=null)
      tc.setLength(term.length());
    if (keyTerm==null)
      keyTerm= term;
    if (keyTerm!=null) {
      keyTerm = keyTerm.replaceAll("[ '()\\-_./,]", "").toLowerCase();
      keyTerm = keyTerm.substring(0, Math.min(keyTerm.length(), 30));
      tc.setKeyTerm(keyTerm);
    }
    this.termCode.add(tc);
    return this;
  }


  public Set<TTIriRef> getMemberOf() {
    return memberOf;
  }

  public EntityDocument setMemberOf(Set<TTIriRef> memberOf) {
    this.memberOf = memberOf;
    return this;
  }


  public String getMatch() {
    return match;
  }

  public EntityDocument setMatch(String match) {
    this.match = match;
    return this;
  }

}

File: src/models/EntityDocument.ts
import z from "zod";

import { SearchTermCodeSchema } from "./SearchTermCode";
import { TTIriRefSchema } from "./TTIriRef";

// export interface EntityDocument {
//   id?: number;
//   iri?: string;
//   name?: string;
//   length?: number;
//   preferredName?: string;
//   code?: string;
//   alternativeCode?: string;
//   scheme?: TTIriRef;
//   type?: TTIriRef[];
//   status?: TTIriRef;
//   termCode?: SearchTermCode[];
//   usageTotal?: number;
//   match?: string;
//   isA?: TTIriRef[];
//   memberOf?: TTIriRef[];
//   subsumptionCount?: number;
//   binding?: string[];
//   isDescendentOf?: TTIriRef[];
// }

export const EntityDocumentSchema = z.strictObject({
  id: z.number().optional(),
  iri: z.url().optional(),
  name: z.string().optional(),
  length: z.number().optional(),
  preferredName: z.string().optional(),
  code: z.string().optional(),
  alternativeCode: z.string().optional(),
  scheme: TTIriRefSchema.optional(),
  type: z.array(TTIriRefSchema).prefault([]),
  status: TTIriRefSchema.optional(),
  termCode: z.array(SearchTermCodeSchema).prefault([]),
  usageTotal: z.number().optional(),
  match: z.string().optional(),
  isA: z.array(TTIriRefSchema).prefault([]),
  memberOf: z.array(TTIriRefSchema).prefault([]),
  subsumptionCount: z.number().optional(),
  binding: z.array(z.string()).prefault([]),
  isDescendentOf: z.array(TTIriRefSchema).prefault([])
});

export type EntityDocument = z.output<typeof EntityDocumentSchema>;

export function isEntityDocument(value: unknown): value is EntityDocument {
  return EntityDocumentSchema.safeParse(value).success;
}
